modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloader.java

```diff
@@ -0,0 +1,449 @@
+            logger.trace("The {} index is not null", EnterpriseGeoIpDownloader.DATABASES_INDEX);
+            if (clusterState.getRoutingTable().index(geoipIndex.getWriteIndex()).allPrimaryShardsActive() == false) {
+                throw new ElasticsearchException("not all primary shards of [" + DATABASES_INDEX + "] index are active");
+            }
+            var blockException = clusterState.blocks().indexBlockedException(ClusterBlockLevel.WRITE, geoipIndex.getWriteIndex().getName());
+            if (blockException != null) {
+                throw blockException;
+            }
+        }
+
+        logger.trace("Updating geoip databases");
+        IngestGeoIpMetadata geoIpMeta = clusterState.metadata().custom(IngestGeoIpMetadata.TYPE, IngestGeoIpMetadata.EMPTY);
+
+        // if there are entries in the cs that aren't in the persistent task state,
+        // then download those (only)
+        // ---
+        // if there are in the persistent task state, that aren't in the cluster state
+        // then nuke those (only)
+        // ---
+        // else, just download everything
+        boolean addedSomething = false;
+        {
+            EnterpriseGeoIpTaskState _state = state;
+            Set<String> metas = Set.copyOf(_state.getDatabases().keySet());
+            for (Map.Entry<String, DatabaseConfigurationMetadata> entry : geoIpMeta.getDatabases().entrySet()) {
+
+                final String id = entry.getKey();
+                DatabaseConfiguration database = entry.getValue().database();
+                if (metas.contains(database.name() + ".mmdb") == false) {
+                    logger.debug("A new database appeared [{}]", database.name());
+
+                    final String accountId = database.maxmind().accountId();
+                    try (HttpClient.PasswordAuthenticationHolder holder = credentialsBuilder.apply(accountId)) {
+                        if (holder == null) {
+                            logger.warn("No credentials found to download database [{}], skipping download...", id);
+                        } else {
+                            processDatabase(holder.get(), id, database);
+                        }
+                    }
+
+                    addedSomething = true;
+                }
+            }
+        }
+
+        boolean droppedSomething = false;
+        {
+            // rip anything out of the task state that doesn't match what's in the cluster state,
+            // that is, if there's no longer an entry for a database in the repository,
+            // then drop it from the task state, too
+            Set<String> databases = geoIpMeta.getDatabases()
+                .values()
+                .stream()
+                .map(c -> c.database().name() + ".mmdb")
+                .collect(Collectors.toSet());
+            EnterpriseGeoIpTaskState _state = state;
+            Collection<Map.Entry<String, Metadata>> metas = List.copyOf(_state.getDatabases().entrySet());
```

Comment:
I left a comment about unexpected `Map.Entry` behavior before, should this one also just be converted into a collection of Tuples?

Reply:
Yeah it probably makes sense here since we're making a copy of it.